mailbox/hbase/src/test/java/org/apache/james/mailbox/hbase/HBaseMailboxManagerProvider.java

```diff
@@ -0,0 +1,102 @@
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.hbase;
+
+import static org.apache.james.mailbox.hbase.HBaseNames.MAILBOXES;
+import static org.apache.james.mailbox.hbase.HBaseNames.MAILBOXES_TABLE;
+import static org.apache.james.mailbox.hbase.HBaseNames.MAILBOX_CF;
+import static org.apache.james.mailbox.hbase.HBaseNames.MESSAGES;
+import static org.apache.james.mailbox.hbase.HBaseNames.MESSAGES_META_CF;
+import static org.apache.james.mailbox.hbase.HBaseNames.MESSAGES_TABLE;
+import static org.apache.james.mailbox.hbase.HBaseNames.MESSAGE_DATA_BODY_CF;
+import static org.apache.james.mailbox.hbase.HBaseNames.MESSAGE_DATA_HEADERS_CF;
+import static org.apache.james.mailbox.hbase.HBaseNames.SUBSCRIPTIONS;
+import static org.apache.james.mailbox.hbase.HBaseNames.SUBSCRIPTIONS_TABLE;
+import static org.apache.james.mailbox.hbase.HBaseNames.SUBSCRIPTION_CF;
+
+import org.apache.james.mailbox.MailboxManager;
+import org.apache.james.mailbox.acl.UnionMailboxACLResolver;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.hbase.mail.HBaseModSeqProvider;
+import org.apache.james.mailbox.hbase.mail.HBaseUidProvider;
+import org.apache.james.mailbox.model.MessageId;
+import org.apache.james.mailbox.store.JVMMailboxPathLocker;
+import org.apache.james.mailbox.store.MailboxManagerOptions;
+import org.apache.james.mailbox.store.StoreMailboxAnnotationManager;
+import org.apache.james.mailbox.store.StoreRightManager;
+import org.apache.james.mailbox.store.event.DefaultDelegatingMailboxListener;
+import org.apache.james.mailbox.store.event.MailboxEventDispatcher;
+import org.apache.james.mailbox.store.mail.model.DefaultMessageId;
+
+import com.google.common.base.Throwables;
+
+public class HBaseMailboxManagerProvider {
+    public MailboxManager provideMailboxManager(HBaseClusterSingleton cluster, MailboxManagerOptions options) {
```

Comment:
Ok, I think I can come up with a better design : each backend could provide a MailboxManager.Builder (for test only) that would allow building a default MailboxManager but still allow customization for tests in need of it.
Looks fine to you ?

Reply:
That's a lot of boilerplate code, repeating that builder everywhere.

And every time we do a code change to that constructor, we would need to modify each builder...

In my opinion we are witnessing the consequences of inheritance (as opposed to  composition). After all I concider we are not far from a single storeMailboxManager class. We just need to compose:
 - instanciation of MessageMapper
 - capabilities

I would then prefer invest in that direction. At least it fixes things at the root.

Reply:
> That's a lot of boilerplate code, repeating that builder everywhere.

I can't understand why. Retrieving the default one is as short as what you propose.

> And every time we do a code change to that constructor, we would need to modify each builder...

Yes, 5 or 6 builder to change for a new parameter to MailboxManager. What's the problem ? Do you expect MailboxManager to have an ever growing number of dependencies ? I wish not.

Finally, about StoreMailboxManager, It's another issue, let's manage the current one here.

Reply:
> Finally, about StoreMailboxManager, It's another issue, let's manage the current one here.

No but reorganizing concepts in it requires changing constructor. And right now, it's a hell. Cc @strokyl  @quynhn  .